feat(query): make QueryList notify on changes via an observable

vsavkin opened the pull request:
BREAKING CHANGE:

Before: query.onChange(() => ...);
After: query.changes.subscribe((iterable) => {});

gkalpak commented on modules/angular2/src/core/compiler/query_list.ts:
```diff
@@ -21,45 +22,16 @@ import {getSymbolIterator} from 'angular2/src/core/facade/lang';
  * class Container {
  *   constructor(@Query(Item) items: QueryList<Item>) {
- *     items.onChange(() => console.log(items.length));
+ *     items.changes.subscribe((_) => console.log(items.length));
```
<sub>You could probably omit the parenthesis around `_`.</sub>

rkirov commented:
Update QueryList documentation. I think it used to say 'this will be Observable'

rkirov commented on modules/angular2/src/core/compiler/element_injector.ts:
```diff
@@ -1074,8 +1072,7 @@ export class QueryRef {
   reset(): void {
-    this.list.reset([]);
-    this.list.removeAllCallbacks();
+    this.list = new QueryList<any>();
```
Why are you recreating the list on every reset? I think this can cause problems because some users will store the QueryList in the component.

rkirov commented on modules/angular2/src/core/compiler/query_list.dart:
```diff
@@ -49,10 +28,12 @@ class QueryList<T> extends Object
+
+  /** @private */
+  void notifyOnChanges() {
+    _emitter.add(this);
```
is the delivery sync or async? Currently, a lot of users have to manually trigger asynchronous execution see - https://github.com/angular/angular/issues/4186 

If we are changing to asynchonous, add that a manual call to `scheduleMicrotask` is not needed any longer.

rkirov commented:
Mostly good, just a few conceptual questions before I can LGTM it.

rkirov commented on modules/angular2/src/core/compiler/element_injector.ts:
```diff
@@ -1074,8 +1072,7 @@ export class QueryRef {
   }
 
   reset(): void {
```
lets rename this to `dehydrate`, `hydrate` and match injector cleanup.

vsavkin committed:
feat(query): make QueryList notify on changes via an observable

BREAKING CHANGE:

Before: query.onChange(() => ...);
After: query.changes.subscribe((iterable) => {});

mary-poppins commented:
Merging PR #4395 on behalf of @vsavkin to branch [presubmit-vsavkin-pr-4395](https://github.com/angular/angular/tree/presubmit-vsavkin-pr-4395).

todoubaba commented:
With alpha.38, angular2.d.ts does not define `subscribe` method in `Observable`.
```
 error TS2339: Property 'subscribe' does not exist on type 'Observable'.
```
